File: internal/apiserver/pkg/orchestratorcfg/toolset_config.go
// Package orchestratorcfg provides orchestrator configuration utilities.
//
// Deprecated: The environment-based toolset configuration (RCA_TOOLSET_CONFIG_JSON/PATH)
// has been deprecated. Use the toolset_config_dynamics table instead, managed via
// /v1/internal-strategy-config/toolsets API. This package is kept for backward
// compatibility and may be removed in a future version.
package orchestratorcfg

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"strings"

	v1 "github.com/aiopsre/rca-api/pkg/api/apiserver/v1"
)

const (
	defaultPipeline         = "basic_rca"
	envToolsetConfigJSON    = "RCA_TOOLSET_CONFIG_JSON"
	envToolsetConfigPath    = "RCA_TOOLSET_CONFIG_PATH"
	providerTypeMCPHTTP     = "mcp_http"
	defaultProviderFunction = "call"
)

var (
	// ErrInvalidConfig indicates resolve failed because server toolset config is malformed/unavailable.
	ErrInvalidConfig = errors.New("invalid orchestrator toolset config")
	// ErrToolsetNotFound indicates normalized pipeline has no mapped toolset in config.
	ErrToolsetNotFound = errors.New("orchestrator toolset not found")
)

type toolsetConfig struct {
	Pipelines map[string][]string
	Toolsets  map[string]toolsetV `json:"toolsets"`
}

type toolsetV struct {
	Providers []providerV `json:"providers"`
}

type providerV struct {
	Type       string   `json:"type"`
	AllowTools []string `json:"allow_tools"`
	Name       string   `json:"name"`
	BaseURL    string   `json:"base_url"`
	Scopes     string   `json:"scopes"`
	TimeoutS   *float64 `json:"timeout_s"`
	Module     string   `json:"module"`
	Function   string   `json:"function"`
}

// NormalizePipeline applies pipeline-only normalize semantics used by orchestrator:
// trim+lower, and empty pipeline resolves to basic_rca.
func NormalizePipeline(pipeline string) string {
	normalized := strings.ToLower(strings.TrimSpace(pipeline))
	if normalized == "" {
		return defaultPipeline
	}
	return normalized
}

// Resolve returns one resolved toolset for the input pipeline from server-side config.
func Resolve(pipeline string) (*v1.OrchestratorToolset, error) {
	toolsets, err := ResolveChain(pipeline)
	if err != nil {
		return nil, err
	}
	if len(toolsets) == 0 {
		return nil, fmt.Errorf("%w: pipeline=%s", ErrToolsetNotFound, NormalizePipeline(pipeline))
	}
	return toolsets[0], nil
}

// ResolveChain returns resolved toolset chain (ordered) for the input pipeline from server-side config.
func ResolveChain(pipeline string) ([]*v1.OrchestratorToolset, error) {
	cfg, err := loadConfigFromEnv()
	if err != nil {
		return nil, err
	}

	normalizedPipeline := NormalizePipeline(pipeline)
	toolsetIDs, ok := cfg.Pipelines[normalizedPipeline]
	if !ok || len(toolsetIDs) == 0 {
		return nil, fmt.Errorf("%w: pipeline=%s", ErrToolsetNotFound, normalizedPipeline)
	}
	return resolveToolsetChainByIDs(toolsetIDs, cfg.Toolsets, normalizedPipeline)
}

func loadConfigFromEnv() (*toolsetConfig, error) {
	rawJSON := strings.TrimSpace(os.Getenv(envToolsetConfigJSON))
	path := strings.TrimSpace(os.Getenv(envToolsetConfigPath))

	var rawBytes []byte
	switch {
	case rawJSON != "":
		rawBytes = []byte(rawJSON)
	case path != "":
		fileBytes, err := os.ReadFile(path)
		if err != nil {
			return nil, invalidConfigf("failed to read %s: %v", envToolsetConfigPath, err)
		}
		rawBytes = fileBytes
	default:
		return nil, invalidConfigf("both %s and %s are empty", envToolsetConfigJSON, envToolsetConfigPath)
	}

	rawCfg := struct {
		Pipelines map[string]any      `json:"pipelines"`
		Toolsets  map[string]toolsetV `json:"toolsets"`
	}{}
	if err := json.Unmarshal(rawBytes, &rawCfg); err != nil {
		return nil, invalidConfigf("json decode failed: %v", err)
	}
	if rawCfg.Pipelines == nil || len(rawCfg.Pipelines) == 0 {
		return nil, invalidConfigf("pipelines is empty")
	}
	if rawCfg.Toolsets == nil || len(rawCfg.Toolsets) == 0 {
		return nil, invalidConfigf("toolsets is empty")
	}

	normalizedPipelines := make(map[string][]string, len(rawCfg.Pipelines))
	for pipeline, mapping := range rawCfg.Pipelines {
		normalizedPipeline := NormalizePipeline(pipeline)
		normalizedChain, normalizeErr := normalizePipelineToolsetChain(normalizedPipeline, mapping)
		if normalizeErr != nil {
			return nil, normalizeErr
		}
		normalizedPipelines[normalizedPipeline] = normalizedChain
	}

	cfg := &toolsetConfig{
		Pipelines: normalizedPipelines,
		Toolsets:  rawCfg.Toolsets,
	}
	return cfg, nil
}

func toolsetConfigSourceConfigured() bool {
	return strings.TrimSpace(os.Getenv(envToolsetConfigJSON)) != "" ||
		strings.TrimSpace(os.Getenv(envToolsetConfigPath)) != ""
}

func resolveToolsetChainByIDs(
	toolsetIDs []string,
	toolsetDefinitions map[string]toolsetV,
	normalizedPipeline string,
) ([]*v1.OrchestratorToolset, error) {
	if len(toolsetIDs) == 0 {
		return nil, invalidConfigf("pipeline=%s mapped toolset chain is empty", normalizedPipeline)
	}
	if len(toolsetDefinitions) == 0 {
		return nil, invalidConfigf("pipeline=%s has no toolset definitions", normalizedPipeline)
	}

	out := make([]*v1.OrchestratorToolset, 0, len(toolsetIDs))
	for chainIndex, rawToolsetID := range toolsetIDs {
		toolsetID := strings.TrimSpace(rawToolsetID)
		if toolsetID == "" {
			return nil, invalidConfigf("pipeline=%s has empty toolset_id at index=%d", normalizedPipeline, chainIndex+1)
		}
		toolsetPayload, exists := toolsetDefinitions[toolsetID]
		if !exists {
			return nil, invalidConfigf("pipeline=%s references missing toolset_id=%s", normalizedPipeline, toolsetID)
		}
		if len(toolsetPayload.Providers) == 0 {
			return nil, invalidConfigf("pipeline=%s toolset=%s has empty providers", normalizedPipeline, toolsetID)
		}

		resolved := &v1.OrchestratorToolset{
			ToolsetID: toolsetID,
			Providers: make([]*v1.OrchestratorToolsetProvider, 0, len(toolsetPayload.Providers)),
		}
		for index, provider := range toolsetPayload.Providers {
			normalizedProvider, normalizeErr := normalizeProvider(provider, toolsetID, index+1)
			if normalizeErr != nil {
				return nil, normalizeErr
			}
			resolved.Providers = append(resolved.Providers, normalizedProvider)
		}
		out = append(out, resolved)
	}
	if len(out) == 0 {
		return nil, fmt.Errorf("%w: pipeline=%s", ErrToolsetNotFound, normalizedPipeline)
	}
	return out, nil
}

func normalizeProvider(provider providerV, toolsetID string, index int) (*v1.OrchestratorToolsetProvider, error) {
	providerType := strings.ToLower(strings.TrimSpace(provider.Type))
	if providerType == "" {
		return nil, invalidConfigf("toolset=%s provider[%d] type is empty", toolsetID, index)
	}
	if providerType != providerTypeMCPHTTP {
		return nil, invalidConfigf("toolset=%s provider[%d] unsupported type=%s", toolsetID, index, providerType)
	}

	allowTools := make([]string, 0, len(provider.AllowTools))
	seen := make(map[string]struct{}, len(provider.AllowTools))
	for _, tool := range provider.AllowTools {
		normalizedTool := strings.TrimSpace(tool)
		if normalizedTool == "" {
			return nil, invalidConfigf("toolset=%s provider[%d] allow_tools contains empty", toolsetID, index)
		}
		if _, ok := seen[normalizedTool]; ok {
			continue
		}
		seen[normalizedTool] = struct{}{}
		allowTools = append(allowTools, normalizedTool)
	}
	if len(allowTools) == 0 {
		return nil, invalidConfigf("toolset=%s provider[%d] allow_tools is empty", toolsetID, index)
	}

	name := strings.TrimSpace(provider.Name)
	if name == "" {
		name = fmt.Sprintf("%s-%d", providerType, index)
	}
	baseURL := strings.TrimSpace(provider.BaseURL)
	module := strings.TrimSpace(provider.Module)
	function := strings.TrimSpace(provider.Function)
	if function == "" {
		function = defaultProviderFunction
	}

	if providerType == providerTypeMCPHTTP && baseURL == "" {
		return nil, invalidConfigf("toolset=%s provider[%d] base_url is required", toolsetID, index)
	}

	out := &v1.OrchestratorToolsetProvider{
		Type:       providerType,
		AllowTools: allowTools,
		Name:       &name,
		BaseURL:    ptrString(baseURL),
		Scopes:     ptrString(strings.TrimSpace(provider.Scopes)),
		Module:     ptrString(module),
		Function:   ptrString(function),
	}
	if provider.TimeoutS != nil && *provider.TimeoutS > 0 {
		timeout := *provider.TimeoutS
		out.TimeoutSeconds = &timeout
	}
	return out, nil
}

func invalidConfigf(format string, args ...any) error {
	return fmt.Errorf("%w: %s", ErrInvalidConfig, fmt.Sprintf(format, args...))
}

func normalizePipelineToolsetChain(pipeline string, raw any) ([]string, error) {
	switch value := raw.(type) {
	case string:
		toolsetID := strings.TrimSpace(value)
		if toolsetID == "" {
			return nil, invalidConfigf("pipeline=%s mapped toolset_id is empty", pipeline)
		}
		return []string{toolsetID}, nil
	case []any:
		if len(value) == 0 {
			return nil, invalidConfigf("pipeline=%s mapped toolset chain is empty", pipeline)
		}
		chain := make([]string, 0, len(value))
		for index, item := range value {
			toolsetID, ok := item.(string)
			if !ok {
				return nil, invalidConfigf(
					"pipeline=%s mapped toolset_id at index=%d must be string",
					pipeline,
					index+1,
				)
			}
			normalizedToolsetID := strings.TrimSpace(toolsetID)
			if normalizedToolsetID == "" {
				return nil, invalidConfigf("pipeline=%s mapped toolset_id is empty at index=%d", pipeline, index+1)
			}
			chain = append(chain, normalizedToolsetID)
		}
		return chain, nil
	default:
		return nil, invalidConfigf("pipeline=%s mapped value has invalid type=%T", pipeline, raw)
	}
}

func ptrString(v string) *string {
	if v == "" {
		return nil
	}
	return &v
}

File: tools/ai-orchestrator/orchestrator/tooling/toolset_config.py
from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path
from typing import Any


def normalize_pipeline_key(pipeline: str | None) -> str:
    normalized = str(pipeline or "").strip().lower()
    if not normalized:
        return "basic_rca"
    return normalized


def normalize_tool_name(tool: str | None) -> str:
    normalized = str(tool or "").strip().lower()
    if normalized.startswith("mcp."):
        return normalized[4:]
    return normalized


@dataclass(frozen=True)
class ProviderConfig:
    provider_type: str
    allow_tools: tuple[str, ...]
    name: str = ""
    base_url: str = ""
    scopes: str = ""
    timeout_s: float = 10.0
    module: str = ""
    function: str = "call"


@dataclass(frozen=True)
class ToolsetDefinition:
    toolset_id: str
    providers: tuple[ProviderConfig, ...]


@dataclass(frozen=True)
class ToolsetConfig:
    pipelines: dict[str, tuple[str, ...]]
    toolsets: dict[str, ToolsetDefinition]

    def resolve_toolset_id(self, pipeline: str | None) -> str:
        return self.get_toolset_chain(pipeline)[0]

    def get_toolset_chain(self, pipeline: str | None) -> list[str]:
        normalized_pipeline = normalize_pipeline_key(pipeline)
        raw = self.pipelines.get(normalized_pipeline)
        if raw is None:
            raise ValueError(f"pipeline={normalized_pipeline} is not mapped to any toolset")
        if isinstance(raw, str):
            chain = [raw]
        elif isinstance(raw, (list, tuple)):
            chain = list(raw)
        else:
            raise ValueError(f"pipeline={normalized_pipeline} has invalid toolset mapping type={type(raw).__name__}")

        normalized_chain: list[str] = []
        for index, item in enumerate(chain, start=1):
            toolset_id = str(item).strip()
            if not toolset_id:
                raise ValueError(f"pipeline={normalized_pipeline} has empty toolset_id at index={index}")
            if toolset_id not in self.toolsets:
                raise ValueError(f"pipeline={normalized_pipeline} references missing toolset_id={toolset_id}")
            normalized_chain.append(toolset_id)
        if not normalized_chain:
            raise ValueError(f"pipeline={normalized_pipeline} has empty toolset chain")
        if len(normalized_chain) != len(chain):
            raise ValueError(f"pipeline={normalized_pipeline} contains invalid empty toolset_id entries")
        return normalized_chain

    def validate_references(self) -> None:
        for pipeline, chain in self.pipelines.items():
            if not chain:
                raise ValueError(f"pipeline={pipeline} has empty toolset chain")
            for index, toolset_id in enumerate(chain, start=1):
                normalized_toolset_id = str(toolset_id).strip()
                if not normalized_toolset_id:
                    raise ValueError(f"pipeline={pipeline} has empty toolset_id at index={index}")
                if normalized_toolset_id not in self.toolsets:
                    raise ValueError(f"pipeline={pipeline} references missing toolset_id={normalized_toolset_id}")
                if normalized_toolset_id != toolset_id:
                    raise ValueError(
                        f"pipeline={pipeline} has non-normalized toolset_id at index={index}: {toolset_id!r}"
                    )

    def get_toolset(self, toolset_id: str) -> ToolsetDefinition:
        normalized_id = str(toolset_id).strip()
        if not normalized_id:
            raise ValueError("toolset_id is required")
        toolset = self.toolsets.get(normalized_id)
        if toolset is None:
            raise ValueError(f"toolset={normalized_id} is not defined")
        return toolset


def load_toolset_config_from_env(settings: Any) -> ToolsetConfig:
    raw_json = str(getattr(settings, "toolset_config_json", "") or "").strip()
    path = str(getattr(settings, "toolset_config_path", "") or "").strip()

    payload: Any
    if raw_json:
        payload = _decode_json(raw_json, source="TOOLSET_CONFIG_JSON")
    elif path:
        file_content = _read_text(path)
        payload = _decode_json(file_content, source=f"TOOLSET_CONFIG_PATH={path}")
    else:
        raise ValueError("toolset config is empty: set TOOLSET_CONFIG_JSON or TOOLSET_CONFIG_PATH")
    return load_toolset_config(payload)


def load_toolset_config(payload: Any) -> ToolsetConfig:
    return _parse_toolset_config(payload)


def _read_text(path: str) -> str:
    try:
        return Path(path).expanduser().read_text(encoding="utf-8")
    except OSError as exc:
        raise ValueError(f"failed to read toolset config file: {path}: {exc}") from exc


def _decode_json(raw: str, *, source: str) -> Any:
    try:
        return json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ValueError(f"invalid toolset config JSON from {source}: {exc}") from exc


def _parse_toolset_config(payload: Any) -> ToolsetConfig:
    if not isinstance(payload, dict):
        raise ValueError("toolset config must be a JSON object")

    pipelines_raw = payload.get("pipelines")
    if not isinstance(pipelines_raw, dict):
        raise ValueError("toolset config requires object field: pipelines")

    toolsets_raw = payload.get("toolsets")
    if not isinstance(toolsets_raw, dict):
        raise ValueError("toolset config requires object field: toolsets")

    pipelines: dict[str, tuple[str, ...]] = {}
    for raw_pipeline, raw_toolset_id in pipelines_raw.items():
        pipeline = normalize_pipeline_key(str(raw_pipeline))
        if isinstance(raw_toolset_id, list):
            toolset_chain: list[str] = []
            for index, item in enumerate(raw_toolset_id, start=1):
                if not isinstance(item, str):
                    raise ValueError(
                        f"pipeline={pipeline} has invalid toolset_id type at index={index}: {type(item).__name__}"
                    )
                normalized_item = item.strip()
                if not normalized_item:
                    raise ValueError(f"pipeline={pipeline} has empty toolset_id at index={index}")
                toolset_chain.append(normalized_item)
        elif isinstance(raw_toolset_id, str):
            single = raw_toolset_id.strip()
            if not single:
                raise ValueError(f"pipeline={pipeline} has empty toolset chain")
            toolset_chain = [single]
        else:
            if raw_toolset_id is None:
                raise ValueError(f"pipeline={pipeline} has empty toolset chain")
            raise ValueError(
                f"pipeline={pipeline} has invalid toolset mapping type={type(raw_toolset_id).__name__}: "
                "expected string or list[string]"
            )
        if not toolset_chain:
            raise ValueError(f"pipeline={pipeline} has empty toolset chain")
        pipelines[pipeline] = tuple(toolset_chain)

    toolsets: dict[str, ToolsetDefinition] = {}
    for raw_toolset_id, raw_toolset_payload in toolsets_raw.items():
        toolset_id = str(raw_toolset_id or "").strip()
        if not toolset_id:
            raise ValueError("toolset id must not be empty")
        if not isinstance(raw_toolset_payload, dict):
            raise ValueError(f"toolset payload must be object: toolset={toolset_id}")
        providers_raw = raw_toolset_payload.get("providers")
        if not isinstance(providers_raw, list) or not providers_raw:
            raise ValueError(f"toolset={toolset_id} requires non-empty providers list")

        providers: list[ProviderConfig] = []
        for index, provider_raw in enumerate(providers_raw, start=1):
            providers.append(_parse_provider(provider_raw, toolset_id=toolset_id, provider_index=index))
        toolsets[toolset_id] = ToolsetDefinition(toolset_id=toolset_id, providers=tuple(providers))

    config = ToolsetConfig(pipelines=pipelines, toolsets=toolsets)
    config.validate_references()
    return config


def _parse_provider(payload: Any, *, toolset_id: str, provider_index: int) -> ProviderConfig:
    if not isinstance(payload, dict):
        raise ValueError(f"provider entry must be object: toolset={toolset_id} index={provider_index}")

    provider_type = str(payload.get("type") or "").strip().lower()
    if not provider_type:
        raise ValueError(f"provider.type is required: toolset={toolset_id} index={provider_index}")

    allow_tools_raw = payload.get("allow_tools")
    if allow_tools_raw is None:
        allow_tools_raw = payload.get("allowTools")
    if not isinstance(allow_tools_raw, list) or not allow_tools_raw:
        raise ValueError(
            f"provider.allow_tools must be non-empty list: toolset={toolset_id} index={provider_index}"
        )
    allow_tools = tuple(_normalize_allow_tools(allow_tools_raw, toolset_id=toolset_id, provider_index=provider_index))

    name = str(payload.get("name") or "").strip()
    if not name:
        name = f"{provider_type}-{provider_index}"

    timeout_raw = payload.get("timeout_s")
    if timeout_raw is None:
        timeout_raw = payload.get("timeoutSeconds")
    timeout_s = _coerce_timeout(timeout_raw, default=10.0)

    base_url = str(payload.get("base_url") or payload.get("baseURL") or "").strip()
    scopes = str(payload.get("scopes") or "").strip()
    module = str(payload.get("module") or "").strip()
    function = str(payload.get("function") or "call").strip() or "call"

    if provider_type == "mcp_http":
        if not base_url:
            raise ValueError(f"provider.base_url is required for mcp_http: toolset={toolset_id} name={name}")
    elif provider_type == "skills":
        if not module:
            raise ValueError(f"provider.module is required for skills: toolset={toolset_id} name={name}")
    else:
        raise ValueError(f"unsupported provider.type={provider_type}: toolset={toolset_id} name={name}")

    return ProviderConfig(
        provider_type=provider_type,
        allow_tools=allow_tools,
        name=name,
        base_url=base_url,
        scopes=scopes,
        timeout_s=timeout_s,
        module=module,
        function=function,
    )


def _coerce_timeout(raw: Any, *, default: float) -> float:
    if raw is None:
        return default
    if isinstance(raw, bool):
        return default
    try:
        value = float(raw)
    except (TypeError, ValueError):
        return default
    if value <= 0:
        return default
    return value


def _normalize_allow_tools(raw: list[Any], *, toolset_id: str, provider_index: int) -> list[str]:
    out: list[str] = []
    seen: set[str] = set()
    for item in raw:
        normalized = normalize_tool_name(str(item))
        if not normalized:
            raise ValueError(
                "provider.allow_tools contains empty item: "
                f"toolset={toolset_id} index={provider_index}"
            )
        if normalized in seen:
            continue
        seen.add(normalized)
        out.append(normalized)
    if not out:
        raise ValueError(f"provider.allow_tools normalized to empty: toolset={toolset_id} index={provider_index}")
    return out
